### PR TITLE
Issue #389: Update self message router and feeback

### PR DESCRIPTION
**Pull request by rschnekenbu**
- update the self router to not use bendpoints
- update feedback to switch from self to oblique and vice versa
- update management of start/end messages for execution specifications
- add some tests for self messages corner cases
- add some (weird) refresh method for tests, to get them green on remote
build. working fine locally PDE and maven, but not on remote build
server
- update anchors on executions to manage automatically the sides.

Change-Id: Ieb479d5b1109c4f4b4a76c0df5e45b48c73192d2
Signed-off-by: Remi Schnekenburger <rschnekenburger@eclipsesource.com>

### CODE REVIEW

**Comment by cdamus** [comment]
Actually, there's also a test failure that is currently hidden by one of the assumption violations, that indicates a real regression.  In test `LifelineSwitchingUITest::ExecutionSpanningOccurrences::switchLifeline`, it used to be that when a message becomes a self-message by virtue of the execution specification moving from one lifeline to another when re-targeting a synchronous message, that self-message was shaped properly with the minimal vertical span.  However, with the changes in this pull request, that message now has no vertical size.  It's this scenario:

Retarget sync message with execution

But worse than this is what happens when attempting the same edit but by moving the execution specification directly (not re-targeting the request message):

Move execution directly

This really breaks the diagram, including an error dialog with cycle in the dependency graph.

**Comment by cdamus** [comment]
Hi, @rschnekenbu. I've made pull request #452 targeting your branch that should clear up the assumption violations on Mac platform.  It also fixes some that I think were happening on all platforms (I see them in the build log also), one of which exposes now a test failure that identifies an actual regression in the editor's behaviour.

File: plugins/org.eclipse.papyrus.uml.diagram.sequence.figure/src/org/eclipse/papyrus/uml/diagram/sequence/figure/anchors/LifelineBodyAnchor.java [hunk 1]
```diff
@@ -45,13 +42,24 @@ public String getTerminal() {
...
-		int realHeight = (int)Math.round(boundedHeight * getScale(lifelinebodyFigure));
-		Point location = new Point(0, realHeight);
-		location.translate(body.getTopLeft());
-		return location;
+		return getReferencePoint();
+	}
+
+	@Override
+	public Point getReferencePoint() {
+		if (getOwner() == null) {
```
[the remark this answers is not in the document]
Comment (rschnekenbu) [reply plugins/org.eclipse.papyrus.uml.diagram.sequence.figure/src/org/eclipse/papyrus/uml/diagram/sequence/figure/anchors/LifelineBodyAnchor.java]: I duplicated the behavior of the overridden method here. I did not encounter this case during testing.

File: plugins/org.eclipse.papyrus.uml.interaction.model/src/org/eclipse/papyrus/uml/interaction/internal/model/commands/SetCoveredCommand.java [hunk 2]
```diff
@@ -448,14 +544,14 @@ private boolean anyDestructionOccurrenceBefore(OptionalInt absoluteY) {
 			Command bend;
 			if (otherEnd.isSend()) {
 				// We're reconnecting the source end, but we need to maintain the gap
-				int newYPosition = currentYPosition - height;
-				Shape newAttachShape = executionShapeAt(lifeline, newYPosition).orElse(lifelineBody);
+				// int newYPosition = currentYPosition - height;
```
[the remark this answers is not in the document]
Comment (rschnekenbu) [reply plugins/org.eclipse.papyrus.uml.interaction.model/src/org/eclipse/papyrus/uml/interaction/internal/model/commands/SetCoveredCommand.java]: This will indeed be deleted in next patch. Thanks for catching it!

File: plugins/org.eclipse.papyrus.uml.interaction.model/src/org/eclipse/papyrus/uml/interaction/internal/model/spi/impl/DefaultDiagramHelper.java [hunk 3]
```diff
@@ -346,31 +344,30 @@ public Command reconnectDestructionOccurrenceShape(Shape destructionView, Connec
...
 
-			// We need to have a bendpoints list, even if it's empty
-			RelativeBendpoints bendpoints = (RelativeBendpoints)result
-					.createBendpoints(NotationPackage.Literals.RELATIVE_BENDPOINTS);
-
 			// Specific routing for self-messages
 			if (selfMessage) {
 				result.setRouting(Routing.RECTILINEAR_LITERAL);
-
-				int xOffset = layoutHelper().getConstraints().getMinimumWidth(result);
```
[the remark this answers is not in the document]
Comment (rschnekenbu) [reply plugins/org.eclipse.papyrus.uml.interaction.model/src/org/eclipse/papyrus/uml/interaction/internal/model/spi/impl/DefaultDiagramHelper.java]: ;-)

File: plugins/org.eclipse.papyrus.uml.interaction.model/src/org/eclipse/papyrus/uml/interaction/internal/model/spi/impl/DefaultDiagramHelper.java [hunk 4]
```diff
@@ -477,6 +455,24 @@ public Command reconnectSource(Connector connector, Shape newSource, int yPositi
...
+		boolean selfMessage = (sourceLL != null) && (sourceLL == targetLL);
+
+		Routing newRouting = Routing.MANUAL_LITERAL;
+		if (selfMessage) {
+			newRouting = Routing.RECTILINEAR_LITERAL;
+		}
+
+		// should force?
+		if (newRouting != connector.getRouting()) {
+			result.chain(SetCommand.create(editingDomain, connector,
```
[the remark this answers is not in the document]
Comment (rschnekenbu) [reply plugins/org.eclipse.papyrus.uml.interaction.model/src/org/eclipse/papyrus/uml/interaction/internal/model/spi/impl/DefaultDiagramHelper.java]: Will be integrated in the next patch

File: plugins/org.eclipse.papyrus.uml.interaction.model/src/org/eclipse/papyrus/uml/interaction/internal/model/spi/impl/DefaultDiagramHelper.java [hunk 5]
```diff
@@ -506,6 +502,23 @@ public Command reconnectTarget(Connector connector, Shape newTarget, int yPositi
...
+		Lifeline targetLL = getLifeline(newTarget);
+		boolean selfMessage = (sourceLL != null) && (sourceLL == targetLL);
+
+		Routing newRouting = Routing.MANUAL_LITERAL;
+		if (selfMessage) {
+			newRouting = Routing.RECTILINEAR_LITERAL;
+		}
+		// should force?
+		if (newRouting != connector.getRouting()) {
+			result.chain(SetCommand.create(editingDomain, connector,
```
[the remark this answers is not in the document]
Comment (rschnekenbu) [reply plugins/org.eclipse.papyrus.uml.interaction.model/src/org/eclipse/papyrus/uml/interaction/internal/model/spi/impl/DefaultDiagramHelper.java]: Will be integrated in the next patch

File: plugins/org.eclipse.papyrus.uml.interaction.model/src/org/eclipse/papyrus/uml/interaction/internal/model/commands/SetOwnerCommand.java [hunk 6]
```diff
@@ -210,6 +317,24 @@ protected Command createCommand(MExecution execution, MLifeline lifeline) {
 		return result;
 	}
 
+	private boolean isStartingSelfMessage(MExecution execution) {
+		Optional<MLifeline> sendStartLL = Executions.getStartMessage(execution).flatMap(MMessage::getSend)
```
[the remark this answers is not in the document]
Comment (rschnekenbu) [reply plugins/org.eclipse.papyrus.uml.interaction.model/src/org/eclipse/papyrus/uml/interaction/internal/model/commands/SetOwnerCommand.java]: I have created the following issue #455 to track that improvement.

File: plugins/org.eclipse.papyrus.uml.diagram.sequence.figure/src/org/eclipse/papyrus/uml/diagram/sequence/figure/anchors/ExecutionSpecificationSideAnchor.java [hunk 7]
```diff
@@ -29,7 +30,7 @@
 
 	public ExecutionSpecificationSideAnchor(IFigure figure, int height) {
 		super(figure);
-		this.side = LEFT;
+		this.side = RIGHT;
```
[the remark this answers is not in the document]
Comment (rschnekenbu) [reply plugins/org.eclipse.papyrus.uml.diagram.sequence.figure/src/org/eclipse/papyrus/uml/diagram/sequence/figure/anchors/ExecutionSpecificationSideAnchor.java]: Added a comment in the code, but this change is more a cosmetic one. The side is updated every time during creations and redirections, so default side shoud not be relevant.

File: plugins/org.eclipse.papyrus.uml.diagram.sequence.figure/src/org/eclipse/papyrus/uml/diagram/sequence/figure/anchors/ExecutionSpecificationEndAnchor.java [hunk 8]
```diff
@@ -36,6 +36,13 @@ public void setConnectionSide(int side) {
 		}
 	}
 
+	@Override
+	public Point getReferencePoint() {
+		Rectangle body = getOwner().getBounds().getCopy();
+		getOwner().translateToAbsolute(body);
+		return body.getBottomLeft();
```
[the remark this answers is not in the document]
Comment (rschnekenbu) [reply plugins/org.eclipse.papyrus.uml.diagram.sequence.figure/src/org/eclipse/papyrus/uml/diagram/sequence/figure/anchors/ExecutionSpecificationEndAnchor.java]: modified with the next patch. Testing did not show any difference, as the real position is computing the right side. In any case, congruency should be kept, so the code was fixed. Thanks for catching it.

File: plugins/org.eclipse.papyrus.uml.diagram.sequence.runtime/src/org/eclipse/papyrus/uml/diagram/sequence/runtime/internal/edit/parts/MessageEditPart.java [hunk 9]
```diff
@@ -224,6 +243,43 @@ protected void refreshRouterChange() {
 		// No bendpoints in messages to refresh
 	}
 
+	@Override
+	protected void handlePropertyChangeEvent(PropertyChangeEvent event) {
+
```
[the remark this answers is not in the document]
Comment (rschnekenbu) [reply plugins/org.eclipse.papyrus.uml.diagram.sequence.runtime/src/org/eclipse/papyrus/uml/diagram/sequence/runtime/internal/edit/parts/MessageEditPart.java]: This one is not able to differentiate if this property is changed only for feedback or for the "real" connection. So it is always computing the router based on the routing style in the model. We do not change this style until the element is finally connected, to avoid affecting the model and issues during cancel. 
So I prefer to avoid this call, and let the feedback management to update the router for the feedback itself.

File: plugins/org.eclipse.papyrus.uml.diagram.sequence.runtime/src/org/eclipse/papyrus/uml/diagram/sequence/runtime/internal/edit/policies/MessageEndpointEditPolicy.java [hunk 10]
```diff
@@ -121,11 +134,72 @@ protected FeedbackHelper getFeedbackHelper(ReconnectRequest request) {
...
+			target = rr.getConnectionEditPart().getTarget();
+		} else {
+			source = rr.getConnectionEditPart().getSource();
+			target = rr.getTarget();
+		}
+
+		return getConnectionRouter(source, target);
+	}
+
+	protected ConnectionRouter getConnectionRouter(EditPart source, EditPart target) {
```
[the remark this answers is not in the document]
Comment (rschnekenbu) [reply plugins/org.eclipse.papyrus.uml.diagram.sequence.runtime/src/org/eclipse/papyrus/uml/diagram/sequence/runtime/internal/edit/policies/MessageEndpointEditPolicy.java]: Issue #456 was opened to adress this task. The proper solution may require some time to implement and test.
